// @volar/kit/out/createProject.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (k !== "default" && Object.prototype.hasOwnProperty.call(mod, k)) __createBinding(result, mod, k);
    __setModuleDefault(result, mod);
    return result;
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.createProject = exports.createInferredProject = void 0;
const path = __importStar(require("typesafe-path/posix"));
const utils_1 = require("./utils");
function createInferredProject(rootPath, getScriptFileNames, compilerOptions = utils_1.defaultCompilerOptions) {
    return createProjectBase(rootPath, () => ({
        options: compilerOptions,
        fileNames: getScriptFileNames().map(utils_1.asPosix),
    }));
}
exports.createInferredProject = createInferredProject;
function createProject(sourceTsconfigPath, extraFileExtensions = [], existingOptions) {
    const ts = require('typescript');
    const tsconfigPath = (0, utils_1.asPosix)(sourceTsconfigPath);
    return createProjectBase(path.dirname(tsconfigPath), () => {
        const parsed = ts.parseJsonSourceFileConfigFileContent(ts.readJsonConfigFile(tsconfigPath, ts.sys.readFile), ts.sys, path.dirname(tsconfigPath), existingOptions, tsconfigPath, undefined, extraFileExtensions);
        parsed.fileNames = parsed.fileNames.map(utils_1.asPosix);
        return parsed;
    });
}
exports.createProject = createProject;
function createProjectBase(rootPath, createParsedCommandLine) {
    const ts = require('typescript');
    const languageHost = {
        workspacePath: rootPath,
        rootPath: rootPath,
        getCompilationSettings: () => {
            return parsedCommandLine.options;
        },
        getProjectVersion: () => {
            checkRootFilesUpdate();
            return projectVersion.toString();
        },
        getScriptFileNames: () => {
            checkRootFilesUpdate();
            return parsedCommandLine.fileNames;
        },
        getScriptSnapshot: (fileName) => {
            if (!scriptSnapshotsCache.has(fileName)) {
                const fileText = ts.sys.readFile(fileName, 'utf8');
                if (fileText !== undefined) {
                    scriptSnapshotsCache.set(fileName, ts.ScriptSnapshot.fromString(fileText));
                }
                else {
                    scriptSnapshotsCache.set(fileName, undefined);
                }
            }
            return scriptSnapshotsCache.get(fileName);
        },
    };
    let scriptSnapshotsCache = new Map();
    let parsedCommandLine = createParsedCommandLine();
    let projectVersion = 0;
    let shouldCheckRootFiles = false;
    return {
        languageHost,
        fileUpdated(fileName) {
            fileName = (0, utils_1.asPosix)(fileName);
            if (scriptSnapshotsCache.has(fileName)) {
                projectVersion++;
                scriptSnapshotsCache.delete(fileName);
            }
        },
        fileDeleted(fileName) {
            fileName = (0, utils_1.asPosix)(fileName);
            if (scriptSnapshotsCache.has(fileName)) {
                projectVersion++;
                scriptSnapshotsCache.delete(fileName);
                parsedCommandLine.fileNames = parsedCommandLine.fileNames.filter(name => name !== fileName);
            }
        },
        fileCreated(fileName) {
            fileName = (0, utils_1.asPosix)(fileName);
            shouldCheckRootFiles = true;
        },
        reload() {
            scriptSnapshotsCache.clear();
            projectVersion++;
            parsedCommandLine = createParsedCommandLine();
        },
    };
    function checkRootFilesUpdate() {
        if (!shouldCheckRootFiles)
            return;
        shouldCheckRootFiles = false;
        const newParsedCommandLine = createParsedCommandLine();
        if (newParsedCommandLine.fileNames.length !== parsedCommandLine.fileNames.length) {
            parsedCommandLine.fileNames = newParsedCommandLine.fileNames;
            projectVersion++;
        }
    }
}
